### PR TITLE
Add pre-commit hooks

### LINKED ISSUE #67: Add `pre-commit` hooks

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #67 

I don't know which ones you wanted, so I just added a few I like (disclaimer: I'm the author of one of them, [nbQA](https://github.com/nbQA-dev/nbQA))

### CODE REVIEW

**Comment by rlouf** [comment]
Nice ! This is basically to replace the use of the [Makefile](https://github.com/blackjax-devs/blackjax/blob/main/Makefile) for developers. I didn't see `mypy` and `flake8` in the config file, did I miss them?

**Comment by MarcoGorelli** [comment]
> Nice ! This is basically to replace the use of the [Makefile](https://github.com/blackjax-devs/blackjax/blob/main/Makefile) for developers. I didn't see `mypy` and `flake8` in the config file, did I miss them?

I've just added flake8.

`mypy` works a bit differently when run in `pre-commit`, it might require some extra tinkering to get it to work

> I also wonder if the `pre-commit` and `Check formatting` GH worflows aren't redundant? How is that managed on PyMC3?

Yes, indeed - I've removed the duplicated checks, except for `mypy` for now.
PyMC3 doesn't check mypy (unfortunately), linting/formatting is all done via pre-commit

**Comment by rlouf** [comment]
You can check the [configuration file](https://github.com/aesara-devs/aehmc/blob/main/.pre-commit-config.yaml) in aesara/aehmc as we use `mypy` there.

**Comment by canyon289** [comment]
Thanks for the request remi. Ill review tomorrow although my prior is this will be good since @MarcoGorelli does great work :)

**Comment by MarcoGorelli** [comment]
> You can check the [configuration file](https://github.com/aesara-devs/aehmc/blob/main/.pre-commit-config.yaml) in aesara/aehmc as we use `mypy` there

Thanks - I was hesitant to do that because it would've required some heavy additional dependencies for the hook (`jax` and `numpy`) but it looks like you're ignoring their types anyway - I've added it now

> There seems to be duplicate workflows below, "Check code formatting" and "pre-commit", would it be possible to remove one of those and keep the name "Check code style" ?

Sure, done

**Review by rlouf — CHANGES_REQUESTED** [review]

File: requirements-dev.txt [hunk 1]
```diff
@@ -18,9 +18,9 @@ pycodestyle==2.6.0
 pyflakes==2.2.0
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply requirements-dev.txt:18]: That's a good point, thanks!

**Comment by rlouf** [comment]
I just added `-e ./` in `requirements-dev.txt` as well so developers do not have to do it when working on the library. I'll rebase and merge when the tests pass, great work !

**Comment by rlouf** [comment]
Hey @MarcoGorelli - Any idea why [this error](https://github.com/blackjax-devs/blackjax/runs/3195255328) happens?

**Comment by MarcoGorelli** [comment]
Hey @rlouf !

Yeah, I've occasionally seen it before, if you run it again it'll almost certainly go away

**Comment by MarcoGorelli** [comment]
You could also use [pre-commit-ci](https://pre-commit.ci/) instead of GHActions for this - it has the option to autofix PRs for you, but you can disable it by adding 
```
ci:
    autofix_prs: false
```
to the top of `.pre-commit-config.yaml`

**Pull request merged**